Document stepper (#62219)
Automattic/wp-calypso: client/landing/stepper/README.md

> <img src="https://user-images.githubusercontent.com/17054134/159939643-4a3a7893-ab38-4223-ba5d-5dfe64282f07.png" alt="Stepper logo" width="400">

Reviewer: Haha I love the logo 👍 👍

# Stepper Signup and Onboarding Framework

> The stepper framework is a new framework for quickly spinning up sign up flows. It's meant to make non-linearly easy without sacrificing agility ✨. 

Reviewer 2: Maybe `non-linearity`?

> The stepper framework is a new framework for quickly spinning up sign up flows. It's meant to make non-linearly easy without sacrificing agility ✨. 

Reviewer: ```suggestion
The stepper framework is a new framework for quickly spinning up sign-up flows. This is a non-linear solution that doesn't sacrifice agility ✨. 
```

This is a bit wordy and doesn't read well. Just a suggestion.

## Non-linearity

> It has been tricky for us to create flows that input-driven steps. Stepper makes it easy by allowing flows to create their own two hooks `useSteps`, and `useStepNavigation`. These hooks have access to the state of of the flow so they can make decisions based on that.

Reviewer 2: `create flows that` ... `input-driven steps`: missing something
`to the state of of the flow`: double of

### Example flow

```ts
import type { StepPath } from './internals/steps-repository';
import type { Flow } from './internals/types';

export const exampleFlow: Flow = {
	useSteps(): Array< StepPath > {
		return [ 'domain', 'design' ];
	},
	useStepNavigation( currentStep, navigate ) {
		const goBack = () => {
			if ( currentStep === 'domain' ) {
				navigate( 'design' );
			} else {
				navigate( 'domain' );
			}
		};
		const goNext = goBack;

> 		return { goNext, goBack, goToStep };

Reviewer 2: goToStep is not defined in the example

	 * Call this function if you want to go to the proceed down the flow.
	 */
	goNext: () => void;
	/**
	 * Call this function if you want to jump to a certain step.
	 */
	goToStep?: ( step: StepPath ) => void;
	/**
	 * Submits the answers provided in the flow
	 */
	submit?: ( providedDependencies?: ProvidedDependencies, ...params: string[] ) => void;
};
```

Since this is a hook, it can access any state from any store, so you can make dynamic navigation decisions based on the state. [Here](./declarative-flow/site-setup-flow.ts) is a developed example of this hook.

```ts
import type { StepPath } from './internals/steps-repository';
import type { Flow } from './internals/types';

export const exampleFlow: Flow = {
	useSteps(): Array< StepPath > {
		return [];
	},
	useStepNavigation( currentStep, navigate ) {
        return { goNext, goBack };
	},
};
```

## Reusability
Stepper aims to create a big `steps-repository` that contains the steps and allows them to be recycled and reused. Every step you create is inherently reusable by any future flow. Because steps are like components, they're not parts of the flows, flows just happen to use them.

This creates a couple of restrictions.

> 

Reviewer: ```suggestion

To maintain the reusability and simplicity of this framework it is important that flow-specific styling changes be made to a `flow` stylesheet or in `global.scss`. Each step should have the basic styling necessary to operate on its own just like a package.
```

Maybe best to add something about styling as well?

## State management

> Steps shouldn't have any props other than `navigation` prop which contains the return value of `useStepNavigation`. This object (`navigation`) allows the step to `submit` when done or to move to other steps or to `goNext`, `goBack`, etc..

Reviewer: ```suggestion
Steps shouldn't have any props other than the `navigation` prop which contains the return value of `useStepNavigation`. This object (`navigation`) allows the step to `submit` when done or to move to other steps or to `goNext`, `goBack`, etc..
```

I think this needs a 'the'

The rationale behind this is the following:

> 1. **Small API**: when steps have such little API surface area. They can be reused easily.

Reviewer 2: API surface area, they can

2. **Centralized state**: the `useStepNavigation` hook makes decisions based on the entire flow's state, so the state has to be centralized, i.e: lives entirely in the `onboard` store (or any store inside `packages/data-stores`).

> All the state should be either the onboarding store, or in the internal state of the step. This creates a flat data tree that is exactly one level deep. This allows the steps to be easy to make and to reuse without any added effort.

Reviewer: ```suggestion
All the state should be either in the onboarding store or the internal state of the step. This creates a flat data tree that is exactly one level deep and allows steps to be easily created and re-used.
```

Another wording suggestion.

It takes five minutes to add a new field to the onboarding store. Such as `intent` or `storeType`. 

## Preferring packages

> When creating new code, it's highly-encouraged to put it packages, whether existing like `onboarding` or new packages that you create. This allows us to create flows in the future that live in wp-admin or wherever.

Reviewer 2: put it _into_ packages
